api/apps/v1alpha1/nemo_customizer_types.go

```diff
@@ -0,0 +1,961 @@
+		},
+		{
+			Name:  "LOG_LEVEL",
+			Value: "INFO",
+		},
+		{
+			Name: "WANDB_ENCRYPTION_KEY",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					Key: n.Spec.WandBSecret.APIKeyKey,
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: n.Spec.WandBSecret.Name,
+					},
+				},
+			},
+		},
+		{
+			Name:  "WANDB_SECRET_NAME",
+			Value: n.Spec.WandBSecret.Name,
+		},
+		{
+			Name:  "WANDB_SECRET_KEY",
+			Value: n.Spec.WandBSecret.APIKeyKey,
+		},
+	}
+
+	// Append the environment variables for Postgres
+	envVars = append(envVars, n.GetPostgresEnv()...)
+
+	// Append the environment variables for OTel
+	if n.IsOtelEnabled() {
+		envVars = append(envVars, n.GetOtelEnv()...)
+	}
+
+	return envVars
+}
+
+// IsOtelEnabled return true if Open Telemetry is enabled
+func (n *NemoCustomizer) IsOtelEnabled() bool {
+	return n.Spec.OpenTelemetry.Enabled != nil && *n.Spec.OpenTelemetry.Enabled
+}
+
+// GetOtelEnv generates the OpenTelemetry-related environment variables.
+func (n *NemoCustomizer) GetOtelEnv() []corev1.EnvVar {
+	return []corev1.EnvVar{
+		{
+			Name:  "OTEL_EXPORTER_OTLP_ENDPOINT",
+			Value: n.Spec.OpenTelemetry.ExporterOtlpEndpoint,
+		},
+		{
+			Name:  "OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED",
+			Value: strconv.FormatBool(*n.Spec.OpenTelemetry.LoggingEnabled),
+		},
+		{
+			Name:  "OTEL_TRACES_EXPORTER",
+			Value: n.Spec.OpenTelemetry.TracesExporter,
+		},
+		{
+			Name:  "OTEL_METRICS_EXPORTER",
+			Value: n.Spec.OpenTelemetry.MetricsExporter,
+		},
+		{
+			Name:  "OTEL_LOGS_EXPORTER",
+			Value: n.Spec.OpenTelemetry.LogsExporter,
+		},
+		{
+			Name:  "OTEL_PYTHON_EXCLUDED_URLS",
+			Value: n.Spec.OpenTelemetry.ExcludedUrls,
+		},
+		{
+			Name:  "OTEL_LOG_LEVEL",
+			Value: n.Spec.OpenTelemetry.LogLevel,
+		},
+	}
+}
+
+// GetPostgresEnv returns the PostgreSQL environment variables for a Kubernetes pod.
+func (n *NemoCustomizer) GetPostgresEnv() []corev1.EnvVar {
+	envVars := []corev1.EnvVar{
+		{
+			Name: "POSTGRES_DB_PASSWORD",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					Key: n.Spec.DatabaseConfig.Credentials.PasswordKey,
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: n.Spec.DatabaseConfig.Credentials.SecretName,
+					},
+				},
+			},
+		},
+		{
+			Name:  "POSTGRES_DB_DSN",
+			Value: n.GeneratePostgresConnString(),
+		},
+	}
+
+	return envVars
+}
+
+// GeneratePostgresConnString generates a PostgreSQL connection string using the database config.
+func (n *NemoCustomizer) GeneratePostgresConnString() string {
+	// Construct the connection string
+	connString := fmt.Sprintf(
+		"postgresql://%s:%s@%s:%d/%s",
+		n.Spec.DatabaseConfig.Credentials.User,
+		"$(POSTGRES_DB_PASSWORD)",
+		n.Spec.DatabaseConfig.Host,
+		n.Spec.DatabaseConfig.Port,
+		n.Spec.DatabaseConfig.DatabaseName,
```

Comment:
since `n.Spec.DatabaseConfig` is a pointer and can be empty..
Unless we reject CRs with empty database config, this might cause nil pointer panic